### PR TITLE
[FIRRTL] EmitOMIR: Drop temporary HierPath's that may be invalid

### PR DESCRIPTION
**Pull request by dtzSiFive**
Restore behavior for deleting the HierPath's added in this pass.

They're known to be dead and sometimes are invalid, causing
breakage during circuit verification.

While we cannot delete incoming HierPath's attached to operations in
input IR without a more careful analysis of all possible users, any NLA
explicitly created in this pass can be reasoned about and removed at the
end for cleaner (and valid) IR.

Fixes assertion failure on a real design.

The invalid NLA encountered is when the target is in the root module,
and so we end up making a one-element (InnerRef) NLA.

### CODE REVIEW

**Review by seldridge — APPROVED** [review]
This is unfortunate that this is generating invalid hierarchical paths. 😬 Great catch.

Is there a small test case that could be added?

**Comment by dtzSiFive** [comment]
I'm not confident about the test, it's copied from an existing test and modified to work as I believe it should.  The non-annotation portion of the IR is small though, and the test demonstrates the issue fixed by this change.

I'll split off https://github.com/llvm/circt/pull/3384/commits/4a3fdf8366c27ca134a82b9aa1f36f91e27a6a52 to another PR.

File: test/Dialect/FIRRTL/emit-omir.mlir [hunk 1]
```diff
@@ -471,6 +471,78 @@ firrtl.circuit "SRAMPathsWithNLA" attributes {annotations = [{
...
+      }
+    },
+    {
+      info = #loc,
+      id = "OMID:1",
+      fields = {
+        omType = {info = #loc, index = 0, value = ["OMString:OMLazyModule", "OMString:OMSRAM"]},
+        finalPath = {info = #loc, index = 1, value = {omir.tracker, id = 1, type = "OMMemberReferenceTarget"}}
+      }
+    }
```
Comment (seldridge) [inline test/Dialect/FIRRTL/emit-omir.mlir:496]: Does the test need to doubly-track the SRAM?

File: test/Dialect/FIRRTL/emit-omir.mlir [hunk 2]
```diff
@@ -471,6 +471,78 @@ firrtl.circuit "SRAMPathsWithNLA" attributes {annotations = [{
...
+    %mem2_port = firrtl.mem Undefined {annotations = [{class = "freechips.rocketchip.objectmodel.OMIRTracker", id = 1}], depth = 8, name = "mem2", portNames = ["port"], readLatency = 0 : i32, writeLatency = 1 : i32 } : !firrtl.bundle<addr: uint<3>, en: uint<1>, clk: clock, data flip: uint<42>>
+  }
+}
+
+// CHECK-LABEL: firrtl.circuit "SRAMPathsTopLevel" {
+// CHECK-NOT: firrtl.hierpath
+// CHECK:         firrtl.extmodule @MySRAM()
+// CHECK-NEXT:    firrtl.module @SRAMPathsTopLevel() {
+// CHECK-NEXT:      firrtl.instance mem1 sym [[SYMMEM1:@[a-zA-Z0-9_]+]]
+// CHECK-SAME:        @MySRAM()
```
Comment (seldridge) [inline test/Dialect/FIRRTL/emit-omir.mlir:511]: This test can likely be golfed down to just a check of these two lines and checks of the path and the symbols.

File: test/Dialect/FIRRTL/emit-omir.mlir [hunk 3]
```diff
@@ -471,6 +471,78 @@ firrtl.circuit "SRAMPathsWithNLA" attributes {annotations = [{
...
+// CHECK-NEXT:  sv.verbatim
+
+// CHECK-SAME:           \22id\22: \22OMID:0\22
+// CHECK-SAME:             \22name\22: \22omType\22
+// CHECK-SAME:             \22value\22: [
+// CHECK-SAME:               \22OMString:OMLazyModule\22
+// CHECK-SAME:               \22OMString:OMSRAM\22
+// CHECK-SAME:             ]
+// CHECK-SAME:              \22name\22: \22finalPath\22
+// CHECK-SAME{LITERAL}:    \22value\22: \22OMMemberInstanceTarget:~SRAMPathsTopLevel|{{0}}/{{1}}:{{2}}\22
```
Comment (seldridge) [inline test/Dialect/FIRRTL/emit-omir.mlir:526]: I'd keep this check.

File: test/Dialect/FIRRTL/emit-omir.mlir [hunk 4]
```diff
@@ -471,6 +471,78 @@ firrtl.circuit "SRAMPathsWithNLA" attributes {annotations = [{
...
+// CHECK-SAME:               \22OMString:OMSRAM\22
+// CHECK-SAME:             ]
+// CHECK-SAME:             \22name\22: \22finalPath\22
+// CHECK-SAME{LITERAL}:    \22value\22: \22OMMemberInstanceTarget:~SRAMPathsTopLevel|{{0}}/{{3}}:
+// CHECK-NOT:                {{.+}}
+// CHECK-SAME:               {{[^\\]+}}\22
+
+// CHECK-SAME:  symbols = [
+// CHECK-SAME:    @SRAMPathsTopLevel,
+// CHECK-SAME:    #hw.innerNameRef<@SRAMPathsTopLevel::[[SYMMEM1:@[a-zA-Z0-9_]+]]>,
```
Comment (seldridge) [inline test/Dialect/FIRRTL/emit-omir.mlir:541]: I'd keep this check.

**Review by seldridge — APPROVED** [review]

File: test/Dialect/FIRRTL/emit-omir.mlir (hunk 2 above)
Comment (dtzSiFive) [reply test/Dialect/FIRRTL/emit-omir.mlir]: :+1:, thanks.  Done!

**Pull request merged**